### DISCUSSION TITLE
1.1.420 Image-wise ControlNet and StyleAlign (Hertz et al.)

### DISCUSSION THREAD

**Discussion by lllyasviel** [Mikubill/sd-webui-controlnet#2295] [General]
Since `sd-webui-controlnet` 1.1.420, users will be able to use image-wise controlnets. And, based on image-wise controlnets, the "StyleAlign" is avaliable.

[Style Aligned Image Generation via Shared Attention](https://style-aligned-gen.github.io/)
[Amir Hertz* 1](https://amirhertz.github.io/) [Andrey Voynov* 1](https://style-aligned-gen.github.io/) [Shlomi Fruchter† 1](https://style-aligned-gen.github.io/) [Daniel Cohen-Or† 1,2](https://danielcohenor.com/)
*1 Google Research 2 Tel Aviv University*

![image](https://github.com/Mikubill/sd-webui-controlnet/assets/19834515/22252293-ffb5-4ede-89b0-a821940f1d78)

Previously, all ControlNet units will be applied to all images in your batch. Now, if you use “each ControlNet unit for each image in a batch”, you will have each ControlNet unit for each image in your batch.

For example, if your batch size is 4, then your 

    ControlNet unit 0 will be applied to image 0, 
    ControlNet unit 1 will be applied to image 1, 
    ControlNet unit 2 will be applied to image 2, 
    ControlNet unit 3 will be applied to image 3 ...

And you can use “[StyleAlign] Align image style in the batch” to align the style of all images in a batch (via shared attention reference).

(If the number of batch size does not match ControlNet unit count, the remainder of the division will be used)

# Sanity Check (with CN)

Some example images used in the Sanity Check:

<img src="https://github.com/Mikubill/sd-webui-controlnet/assets/19834515/16f3b4ef-de25-437a-a889-de716d1feec4" width="100">

<img src="https://github.com/Mikubill/sd-webui-controlnet/assets/19834515/7bce08ad-f2de-4593-b96a-04ee773a0878" width="100">

<img src="https://github.com/Mikubill/sd-webui-controlnet/assets/19834515/ae3256dc-cb29-4e06-90c1-e94de4b48798" width="100">

<img src="https://github.com/Mikubill/sd-webui-controlnet/assets/19834515/3b0b57be-f682-4545-91f0-5c2052ac75ed" width="100">

The model is `realisticVisionV51_v51VAE`: https://civitai.com/models/4201?modelVersionId=130072

![image](https://github.com/Mikubill/sd-webui-controlnet/assets/19834515/d7137c7b-aec8-4fba-bc72-e8ade240c869)

Prompt

![image](https://github.com/Mikubill/sd-webui-controlnet/assets/19834515/8c27f6c0-36a0-461a-96cc-2d182cb6e2f5)

Positive:

best quality, very detailed, high resolution, 4k, 8k, 35 mm, a handsome man, looking at viewer, street

Negative:

text, watermark, low quality, medium quality, blurry, censored, deformed, mutated, anime, toon, render, 3d, ilustration, moles, dark skin spots

Parameters:

![image](https://github.com/Mikubill/sd-webui-controlnet/assets/19834515/6df6b5b6-2688-4053-81e4-ec79e835c7dc)

4x Openpose CN (control_v11p_sd15_openpose):

![image](https://github.com/Mikubill/sd-webui-controlnet/assets/19834515/6c8a3899-7eae-49f4-8614-2087cb9eab9e)
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/19834515/27c3eed5-3be3-4ff5-8915-0800b860ef6b)
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/19834515/e7225d2b-dc37-43dc-b533-67389de8c19a)
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/19834515/bb351772-0b29-4781-a206-eeab78d7c3e5)

All default parameters.

Note that here we use openpose (rather than openpose_full) without face landmarks to avoid the style influence of face appearance.

Also, if you do not have 4 controlnet units, go to settings->controlnet->ControlNet unit number to have any number of units.

Check "Each ControlNet unit for each image in a batch"

![image](https://github.com/Mikubill/sd-webui-controlnet/assets/19834515/d273650e-e6b8-4558-9e36-7adfd1e1490f)

Generate, you will get this

![image](https://github.com/Mikubill/sd-webui-controlnet/assets/19834515/4a598ced-e729-45f2-b32f-9e4abbdbb7f1)

The 4 images are generated by these 4 poses

![image](https://github.com/Mikubill/sd-webui-controlnet/assets/19834515/5ffcec27-dd64-43aa-b4cd-e9fe4cfbf53f)

You can see this is what "Each ControlNet unit for each image in a batch".

Then check "[StyleAlign] Align image style in the batch."

![image](https://github.com/Mikubill/sd-webui-controlnet/assets/19834515/2ae71e08-01ea-4650-8031-1a70502acdca)

Generate again, you will get the "StyleAlign" for the 4 images in a same batch:

![image](https://github.com/Mikubill/sd-webui-controlnet/assets/19834515/97c03b74-e0d5-4686-aae0-e9ad97cf738f)

# Sanity Check (with dynamic-prompts)

You need to install https://github.com/adieyal/sd-dynamic-prompts to use prompt permutation.

The grammar is {A|B|C|D} like "a toy {train|bicycle|car|boat}. macro photo. 3d game asset"

The model is `realisticVisionV51_v51VAE`: https://civitai.com/models/4201?modelVersionId=130072

![image](https://github.com/Mikubill/sd-webui-controlnet/assets/19834515/d7137c7b-aec8-4fba-bc72-e8ade240c869)

Positive:

a toy {train|bicycle|car|boat}. macro photo. 3d game asset

negative:

low quality, lowres, ugly, bad

![image](https://github.com/Mikubill/sd-webui-controlnet/assets/19834515/7d01b938-196a-439c-b048-1df458a4904f)

parameters:

![image](https://github.com/Mikubill/sd-webui-controlnet/assets/19834515/8d65d1c4-21f9-45a0-963f-47795fa69ee2)

dynamic-prompts:

![image](https://github.com/Mikubill/sd-webui-controlnet/assets/19834515/a121bccb-78af-4250-90b0-05db9d677320)

then if you click generate and generate without using "StyleAlign", the result is

![image](https://github.com/Mikubill/sd-webui-controlnet/assets/19834515/807996c4-ee25-4faa-ae9c-f1cafb180164)

then check "StyleAlign"

![image](https://github.com/Mikubill/sd-webui-controlnet/assets/19834515/b486dc5e-7e85-4f73-b417-8f724402e9ea)

then the style will be aligned

![image](https://github.com/Mikubill/sd-webui-controlnet/assets/19834515/d18e6514-f183-4e2b-a5e0-ffe415a0dd32)

### FAQ

Q: Does this influence speed?
A: Yes, if you use "StyleAlign", attention context is longer. Generating will be slower. But in my tests not too slow.

Q: Can I use t2ia-style/reference/ip-adapters together with StyleAlign?
A: Yes. But this behavior is still under experiment and may be changed in any future versions. (Not yet very sure what behavior should be the correct behavior because now attention will have formulations both in-batch and globally. It should already work out of box but not extensively tested.)

Q: XL? SSD? Turbo? LCM?
A: yes, yes, yes, and yes. (w/ webui 1.7.0)

**Comment by lllyasviel** [comment]
@catboxanon @huchenlei @ljleb @anyone-who-know-how-to-make-it-work
currently everything in the `batch options` is not related to any script args or infotext. help wanted if anyone know how to make it more consistent to previous codes. if nobody know how to do it, then it will be left as is.

**Comment by sdbds** [comment]
![image](https://github.com/Mikubill/sd-webui-controlnet/assets/8085926/818a3a3f-a3c0-4b33-8975-e8c1e77f4c99)
I think we need set number for batch size and batch count like webui for aligning video animation.

![image](https://github.com/Mikubill/sd-webui-controlnet/assets/8085926/5bee6246-2bc5-4388-9440-c15947e11533)
For animation or video,they use different context batch size (render context frame) and FPS.

**Reply by lllyasviel** [reply]
https://github.com/continue-revolution/sd-webui-animatediff/issues/360

**Reply by sdbds** [reply]
It's great to see this collaboration, continue-revolution is one of my best friend, he's a very nice guy, both in terms of coding ability and academic ability

**Reply by continue-revolution** [reply]
Thanks for your invitation. Very glad to join you guys and improve CN for video generation in A1111

**Reply by lllyasviel** [reply]
@continue-revolution welcome
@huchenlei @ljleb

**Reply by huchenlei** [reply]
Welcome onboard! @continue-revolution

**Comment by matrix4767** [comment]
Hey, a notebook was added on the main github page for sd 1.4, like 20 minutes after this controlnet update. And so would probably work for 1.5 too.

**Reply by lllyasviel** [reply]
realisticVisionV51_v51VAE is sd1.5 model

**Reply by matrix4767** [reply]
This is present in the update?
https://github.com/google/style-aligned/commit/f3a9669db0e19d7c1784f7e8f1bf80a99d604f19#diff-7a71e188435b6f68af74c3dcc72d8df5758a9c190855cc27ba4c2f3bb23315c5

**Comment by catboxanon** [comment]
Pinging @comfyanonymous here since they claim StyleAlign is not implemented correctly. Maybe they can provide more useful info.
https://boards.4channel.org/g/thread/97736971#p97740259

**Reply by lllyasviel** [reply]
This impl is converted from the paper.
there is a source-target adain also mentioned in paper and should be already in reference cn units if users use it.
feel free to ping me if in the future users get alter choices in other software

**Comment by lllyasviel** [comment]
update 12/6:

now also tested dynamic-prompts permutations (see also the updated sanity check)

**Comment by sdbds** [comment]
https://huggingface.co/spaces/fffiloni/video2openpose2
I have made a video2pose gradio referencing this huggingface space that can be extended to all controlnet preprocessors via the control_aux pip package. 
Do I need to add video processing to the sd-webui-control plugin in the future? 
Or should I keep the video independent? 
https://github.com/sdbds/vid2pose/blob/main/video2openpose2.py
This is the current situation. I just added dwpose to the original openpose and it is faster with cuda.
@lllyasviel  @huchenlei @continue-revolution

**Reply by continue-revolution** [reply]
I‘m planning to move everything about CN hook inside my AnimateDiff extension here, so I think ultimately we may want CN to be compatible with videos.

**Comment by drphero** [comment]
> And you can use “[StyleAlign] Align image style in the batch” to align the style of all images in a batch (via shared attention reference).

What is the style being aligned to?

**Reply by kft334** [reply]
From my limited testing it seems that all images beyond the first in a batch are aligned to the first image of the batch. The style-align repo does have style alignment from a reference image now, at least for SDXL, so hopefully that finds it's way here soon too. I'm getting artifacts from the first batch image in some other images in the batch so the problem of separating content from style does still exist in style-align but it does appear to do a better job.

**Reply by lllyasviel** [reply]
"Style Align" align image styles in a batch, it is not "Style Transfer".

**Reply by john-mnz** [reply]
@lllyasviel 
have you seen this?
it seems to be style transfer
https://github.com/google/style-aligned/blob/main/style_aligned_transfer_sdxl.ipynb

**Comment by wardensc2** [comment]
I can't make it work with Reference Adain the others Reference Only or Reference Adain + Attn work fine. Anytimes I select Reference Adain with Batch Option and Style Align On I get this error  message, does anyone know what error is this ?
[Style Align Error.docx](https://github.com/Mikubill/sd-webui-controlnet/files/13628550/Style.Align.Error.docx)

**Comment by enternalsaga** [comment]
I could not replicate results of Sanity Check (with dynamic-prompts)? what did you add in controlnet? Do I need to enable any controlnet or just the stylealign checkbok is fine?

**Reply by xpeng** [reply]
just check stylealign checkbox

**Comment by xpeng** [comment]
is there any problem in negative prompt length larger than 75 when enable StyleAlign?
```RuntimeError: step must be nonzero```

for example in negative prompt box:
```
anime, cartoon, graphic, text, painting, crayon, graphite, abstract, glitch, deformed, mutated, ugly, disfigured, tripod, camera, anime, animation, cartoon, 3D, drawing, painting, (censorship, censored, worst quality, low quality, normal quality, lowres, low details, bad photo, bad photography, bad art:1.4), (watermark, signature, text font, username, error, logo, words, letters, digits, autograph, trademark, name:1.2), (blur, blurry), morbid, ugly, asymmetrical, mutated malformed, mutilated, poorly lit, bad shadow, draft, cropped, out of frame, cut off, censored, jpeg artifacts, out of focus, glitch, duplicate, (bad hands, bad anatomy, bad body, bad face, bad teeth, bad arms, bad legs, deformities:1.3) (ugly hands, ugly anatomy, ugly body, ugly face, ugly teeth, ugly arms, ugly legs, deformities:1.3) ugly fingers, bad fingers, (((ugly nipples, bad nipples, deformed nipples))), (((Bad teeth, ugly teeth)))
```

**Reply by somanchiu** [reply]
It applies not only to the negative prompt, but also to the positive prompt.

**Comment by raonsol** [comment]
Hello, Do you have plan to implement this method with additional way, such as the way like IP-Adapter?

StyleAlign can do so-called "Style Transfer" method, because in their [example](https://github.com/google/style-aligned/blob/main/style_aligned_transfer_sdxl.ipynb) it says StyleAlign can be used in generating images with a style from **"reference image"**. The example also shows the method that uses both **reference image** and **depth map** to generate same style images with different text prompt.

![image](https://github.com/Mikubill/sd-webui-controlnet/assets/31945968/4edeff4c-3e4b-4d4e-9414-682e368fa906)


I think this can be implemented not only by aligning other images with first image during the batch, but some kind of separated type in controlnet unit.
[Here](https://huggingface.co/spaces/ysharma/style-aligned-controlnet) is the example that @yvrjsharma implemented to do something like this.

**Reply by LukcyOne** [reply]
I just want to achieve style alignment by referencing images, perhaps style aligned transfer is a good method.
I have tried the "sref" function of NijiJourney, which can reference the style of images. But I can't find an open-source implementation that can match it with simple operations.

**Comment by xpeng** [comment]
sd1.5 based model is ok.
but failed on openpose controlnet applying for sdxl based models. come with different errors on different openpose controlnet models.

**Comment by qinzhenzhao** [comment]
Great.It's marvelous.
But how can I use API request to generate images by BatchOptions and styleAlign?I mean what is the parameter format?

**Reply by huchenlei** [reply]
Unfortunately, there is currently no way to use style align via API now, as batch options cannot be controlled by API. You can file an issue and we will consider implement/rework it if enough people think it is useful.
